### PR TITLE
refactor: functional test

### LINKED ISSUE #776: Jest Matchers migration
**Is your feature request related to a problem? Please describe.**

Many unit/functional test still do not use Jest matches or stubs

**Describe the solution you'd like**
We have to remove all Node module assert in favor of Jest Matchers 

**Describe alternatives you've considered**
Use setup and teardown to trigger verdaccio servers for functional testing

### PR DESCRIPTION
**Pull request by juanpicado**
**Type:** unit test / build

The following has been addressed in the PR:

*  There is a related issue: #776

**Description:**

Resolves #776

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -102,7 +102,8 @@ jobs:
       - *restore_repo
       - run:
           name: Test with Node 8
-          command: yarn run test
+          command: yarn test
+          command: yarn test:func
```
Comment (ayusharma) [inline .circleci/config.yml]: I think we should keep the full name - `yarn test:functional`.

Comment (juanpicado) [reply .circleci/config.yml]: I will change it

Comment (juanpicado) [reply .circleci/config.yml]: Done

File: src/lib/constants.js [hunk 2]
```diff
@@ -4,4 +4,59 @@ export const HEADERS = {
   JSON: 'application/json',
   JSON_CHARSET: 'application/json; charset=utf-8',
   OCTET_STREAM: 'application/octet-stream',
+  GZIP: 'gzip',
 };
+
+export const HEADER_TYPE = {
```
Comment (ayusharma) [inline src/lib/constants.js]: I like this approach. Easy to maintain constants. 👍

File: test/functional/tags/dist-tags-merge.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      test('server2 should match with sha key from published package', () => {
+        return matchTarBallSha(server2);
+      });
+    });
+
+    describe('should match dist-tags', () => {
+      const matchDisTags = (verdaccioServer, port) => {
+        return verdaccioServer.getPackage(PKG_NAME)
+          .status(HTTP_STATUS.OK)
+          .then(function(body) {
```
Comment (ayusharma) [inline test/functional/tags/dist-tags-merge.js]: Arrow function.

File: test/jest.config.func.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+/* eslint comma-dangle: 0 */
```
Comment (ayusharma) [inline test/jest.config.func.js]: filename. `jest.config.functional.js`

File: test/jest.config.func.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+
+module.exports = {
+  name: 'verdaccio-func-jest',
```
Comment (ayusharma) [inline test/jest.config.func.js]: verdaccio-functional-jest

File: test/functional/basic/basic.js [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+          beforeAll(function () {
+            const pkg = getPackage(PKG_NAME);
+
+            pkg.dist.shasum = createTarballHash().update(readfile('../fixtures/binary')).digest('hex');
+            return server.putVersion(PKG_NAME, PKG_VERSION, pkg)
+              .status(HTTP_STATUS.CREATED)
+              .body_ok(/published/);
+          });
+
+          test('uploading new package version', () => {
```
[the remark this answers is not in the document]
Comment (juanpicado) [reply test/functional/basic/basic.js]: it is not, the previous author intended to print force a report line for the `beforeAll` section above. I'll remove it.

File: test/functional/notifications/notify.js [hunk 7]
```diff
@@ -125,10 +115,10 @@ export default function(express) {
         name: "pkg-test"
```
[the remark this answers is not in the document]
Comment (juanpicado) [reply test/functional/notifications/notify.js]: Mmmm this should be caught by eslint

File: src/lib/constants.js (hunk 2 above)
Comment (juanpicado) [reply src/lib/constants.js]: We need to create a `@verdaccio/error-utils` or something like that to share error messages with plugins, right now is kind of wild west where the plugin dev can set whatever the wants

File: test/functional/package/gzip.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+export default function(server, express) {
+  const PKG_NAME = 'testexp_gzip';
+
+  describe('test gzip support', () => {
+    beforeAll(function() {
+      express.get(`/${PKG_NAME}`, function(req, res) {
+        const version = '0.0.1';
+        const pkg = eval(
+          '(' + readFile('../fixtures/publish.json5')
```
[the remark this answers is not in the document]
Comment (juanpicado) [reply test/functional/package/gzip.js]: I removed the `eval`

File: test/functional/plugins/auth.js [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+      });
+
+  };
+
+  describe('plugin authentication', () => {
+
+    describe('test users authentication', () => {
+
+      test('should not authenticate user1 with wrong password', () => {
+        return requestAuthFail(USER1, WRONG_PASSWORD, 'i don\'t like your password', HTTP_STATUS.UNAUTHORIZED);
```
[the remark this answers is not in the document]
Comment (juanpicado) [reply test/functional/plugins/auth.js]: I cannot change it, it comes from an external source.

**Pull request merged**